### PR TITLE
Update Transport with new API from Trino v406

### PR DESCRIPTION
**Pull request by yiqiangin**
The code changes of this PR mainly include
- Bumping the version of Trino used in Transport from v352 to v406
- Replacing the classes defined in trino-main in v352 to the classes defined in trino-spi in v406
- Replacing some classes of test infrastructure in Trino which are deprecated in v406 with the in-use classes

Tests:
- unit tests are passed in all modules of Transport
- major unit tests are passed in `transportable-udfs-examples` except several test cases fail because of the issue in `QueryAssertions` from Trino v406

### CODE REVIEW

File: transportable-udfs-examples/transportable-udfs-example-udfs/build.gradle [hunk 1]
```diff
@@ -12,6 +12,7 @@ dependencies {
   implementation('com.google.guava:guava:24.1-jre')
   implementation('org.apache.commons:commons-io:1.3.2')
   testImplementation('io.airlift:aircompressor:0.21')
+  testImplementation('org.junit.jupiter:junit-jupiter-api:5.9.2')
```
Comment (ljfgem) [inline transportable-udfs-examples/transportable-udfs-example-udfs/build.gradle:15]: Why do we need to add it?

File: transportable-udfs-test/transportable-udfs-test-spi/src/main/java/com/linkedin/transport/test/spi/SqlStdTester.java [hunk 2]
```diff
@@ -25,7 +25,9 @@ public interface SqlStdTester extends StdTester {
    * @param expectedOutputData  The expected output data from the function call
    * @param expectedOutputType  The expected output type from the function call
    */
-  void assertFunctionCall(String functionCallString, Object expectedOutputData, Object expectedOutputType);
+  default void assertFunctionCall(String functionCallString, Object expectedOutputData, Object expectedOutputType) {
+    throw new UnsupportedOperationException();
+  };
```
Comment (ljfgem) [inline transportable-udfs-test/transportable-udfs-test-spi/src/main/java/com/linkedin/transport/test/spi/SqlStdTester.java:30]: Why is this change necessary?

File: transportable-udfs-test/transportable-udfs-test-trino/build.gradle [hunk 3]
```diff
@@ -20,4 +20,5 @@ dependencies {
   // The io.airlift.slice dependency below has to match its counterpart in trino-root's pom.xml file
   // If not specified, an older version is picked up transitively from another dependency
   implementation(group: 'io.airlift', name: 'slice', version: project.ext.'airlift-slice-version')
+  implementation(group: 'org.assertj', name: 'assertj-core', version: '3.24.2')
```
Comment (ljfgem) [inline transportable-udfs-test/transportable-udfs-test-trino/build.gradle:23]: Why do we need to add it?

File: transportable-udfs-test/transportable-udfs-test-trino/src/main/java/com/linkedin/transport/test/trino/TrinoTester.java [hunk 4]
```diff
@@ -7,44 +7,62 @@
...
 
   @Override
   public void setup(
       Map<Class<? extends TopLevelStdUDF>, List<Class<? extends StdUDF>>> topLevelStdUDFClassesAndImplementations) {
     // Refresh Trino state during every setup call
-    initTestFunctions();
     for (List<Class<? extends StdUDF>> stdUDFImplementations : topLevelStdUDFClassesAndImplementations.values()) {
       for (Class<? extends StdUDF> stdUDF : stdUDFImplementations) {
-        registerScalarFunction(new TrinoTestStdUDFWrapper(stdUDF));
+        _runner.addFunctions(new InternalFunctionBundle(new SqlFunction[]{new TrinoTestStdUDFWrapper(stdUDF)}));
```
Comment (ljfgem) [inline transportable-udfs-test/transportable-udfs-test-trino/src/main/java/com/linkedin/transport/test/trino/TrinoTester.java:65]: Why do we need this change?

File: transportable-udfs-test/transportable-udfs-test-trino/src/main/java/com/linkedin/transport/test/trino/TrinoTester.java [hunk 5]
```diff
@@ -75,7 +95,16 @@ public ToPlatformTestOutputConverter getToPlatformTestOutputConverter() {
...
+    String functionName = testCase.getFunctionCall().getFunctionName();
+    List<Object> parameters = testCase.getFunctionCall().getParameters();
+    List<TestType> testTypes = testCase.getFunctionCall().getInferredParameterTypes();
+    List<String> functionArguments = new ArrayList<>();
+    for (int i = 0; i < parameters.size(); ++i) {
+      functionArguments.add(_sqlFunctionCallGenerator.getFunctionCallArgumentString(parameters.get(i), testTypes.get(i)));
+    }
+    Object expectedOutputType = getPlatformType(testCase.getExpectedOutputType());
+    QueryAssertions.ExpressionAssertProvider expressionAssertProvider = _queryAssertions.function(functionName, functionArguments);
+    assertThat(expressionAssertProvider).hasType((Type) expectedOutputType).isEqualTo(testCase.getExpectedOutput());
```
Comment (ljfgem) [inline transportable-udfs-test/transportable-udfs-test-trino/src/main/java/com/linkedin/transport/test/trino/TrinoTester.java:108]: Why do we need this change?

**Review by ljfgem — COMMENTED** [review]
Thanks @yiqiangin for this PR!
Several points:
1. Could you please add more details to the PR description? Especially for `Replacing some classes of test infrastructure in Trino which are deprecated in v406 with the in-use classes
`, would be great to list them.
2. Given this is the last transport-trino upgrade PR in transport, I would suggest to have an end-to-end local test with Trino 406 local server on some production views with upgraded Trino transport UDFs to ensure there is no unexpected issue.

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/TrinoFactory.java [hunk 6]
```diff
@@ -137,25 +151,22 @@ public StdStruct createStruct(StdType stdType) {
...
+    return TrinoWrapper.createStdType(functionDependencies.getType(typeSignature));
   }
 
   public MethodHandle getOperatorHandle(
       OperatorType operatorType,
       List<Type> argumentTypes,
       InvocationConvention invocationConvention) throws OperatorNotFoundException {
-    if (metadata != null) {
-      return metadata.getScalarFunctionInvoker(metadata.resolveOperator(operatorType, argumentTypes),
+    if (functionManager != null) {
```
Comment (aastha25) [inline transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/TrinoFactory.java:166]: in addition to the new check, should we also retain the original null check for metadata since we are using `metadata.resolveOperator(...)` ?

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/TrinoFactory.java [hunk 7]
```diff
@@ -55,17 +59,27 @@ public class TrinoFactory implements StdFactory {
   final FunctionBinding functionBinding;
   final FunctionDependencies functionDependencies;
   final Metadata metadata;
+  final TypeManager typeManager;
+  final FunctionManager functionManager;
+
+  final Session session;
 
   public TrinoFactory(FunctionBinding functionBinding, FunctionDependencies functionDependencies) {
```
Comment (aastha25) [inline transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/TrinoFactory.java:67]: is it reasonable to assume that this exists for backward compatibility? 
should we mark this deprecated?

File: transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java [hunk 8]
```diff
@@ -141,17 +142,17 @@ public ScalarFunctionImplementation specialize(FunctionBinding functionBinding,
...
+    return new ChoicesSpecializedSqlScalarFunction(
+        boundSignature,
         NULLABLE_RETURN,
         getNullConventionForArguments(nullableArguments),
-        getMethodHandle(stdUDF, functionBinding, nullableArguments, requiredFilesNextRefreshTime));
+        getMethodHandle(stdUDF, boundSignature, nullableArguments, requiredFilesNextRefreshTime));
   }
 
-  private MethodHandle getMethodHandle(StdUDF stdUDF, FunctionBinding functionBinding, boolean[] nullableArguments,
+  private MethodHandle getMethodHandle(StdUDF stdUDF, BoundSignature boundSignature, boolean[] nullableArguments,
```
Comment (aastha25) [inline transportable-udfs-trino/src/main/java/com/linkedin/transport/trino/StdUdfWrapper.java:152]: Seems like `functionBinding` is available in the caller method. 
Since `getMethodHandle(...)` is a private utility, is the change from `functionBinding` -> `boundSignature` necessary or good to have?

File: transportable-udfs-test/transportable-udfs-test-trino/src/main/java/com/linkedin/transport/test/trino/TrinoTester.java [hunk 9]
```diff
@@ -7,44 +7,62 @@
...
+    _featuresConfig = new FeaturesConfig();
+    _runner = LocalQueryRunner.builder(_session).withFeaturesConfig(_featuresConfig).build();
+    _queryAssertions = new QueryAssertions(_runner);
   }
 
   @Override
   public void setup(
       Map<Class<? extends TopLevelStdUDF>, List<Class<? extends StdUDF>>> topLevelStdUDFClassesAndImplementations) {
     // Refresh Trino state during every setup call
-    initTestFunctions();
```
Comment (aastha25) [inline transportable-udfs-test/transportable-udfs-test-trino/src/main/java/com/linkedin/transport/test/trino/TrinoTester.java:44]: could you clarify what was this init function doing previously and how come its not needed anymore?

**Review by aastha25 — COMMENTED** [review]
Thanks for the PR @yiqiangin !

There's some documentation in `transport/docs/required-trino-apis.md` which might be outdated now. Please update the documentation there as well.
 
Please run some integ tests with a local instance of the trino server, and share results for TUDFs that access input data from HDFS.

**Comment by weijiii** [comment]
Curious is `StdUdfWrapper` going to still extend `SqlScalarFunction` with the new SPI?
